mlir/lib/Dialect/Tensor/IR/TensorOps.cpp

```diff
@@ -1622,7 +1623,20 @@ CollapseShapeOp::inferCollapsedType(RankedTensorType type,
     currentDim += dim;
   }
 
-  return RankedTensorType::get(newShape, type.getElementType());
+  auto encoding = type.getEncoding();
+  if (auto v = encoding.dyn_cast_or_null<VerifiableTensorEncoding>()) {
```

Comment:
Seems to me that the conservative option is that if the interface isn't implemented, the encoding should no be propagated.

Reply:
Hi @joker-eph thanks for your response. I think there may be a misunderstanding, as i don't think what you're saying will work, but i'm not sure if its on my part or yours.

From what i read in the code, the previous behavior drops the encoding (not propagated) for all cases when creating a new type. I'm changing this to reuse the encoding on the newly created type. However, the special case test here is that for this particular encoding (the `VerifiabletensorEncoding`), the encoding may not be possible to propagate, since it is dimensionality specific.

I believe the best possible approach would be to update the encoding based on the new dimensionality, but as this isn't an area of the compiler i'm familiar with and that we don't use in our code base, i'm instead falling back to the exiting behavior of dropping the encoding.

Does this make sense?

I'm happy to discuss more if this approach is not what folks want to see.

Reply:
In general dropping is the "safe" thing to do when we transform code and we don't know if the transformation can preserve the semantics of the encoding.

Now there may be a specific argument here you'd want to make, but it should be made explicit, about why propagating "blindly" an opaque encoding would be correct in the absolute? If my attribute does not implement the "VerifiableTensorEncoding", you can't ensure that it is correct to just propagate it right?

Reply:
> My thinking (before this discussion) that the Encoding attribute was opaque, so the code shouldn't know anything about them

But when something is "opaque" we have two approaches:
1) We don't do the transformation (because we risk making an incorrect transformation)
2) We drop the opaque attribute (we don't guarantee propagation, these attributes have to be droppable).

This come down to the guarantee we intend to provide with this attribute though.


> don't change the default, but instead, only transfer the encoding if it can be proved to be transferable (via the VerifiableTensorEncoding interface).

Yes: basically we could consider that an attribute which implements `VerifiableTensorEncoding` is "opt in" into being propagated as long as it verifies.

But even this is sketchy: the "verifier" aspect does not guarantee that the transformation is semantic preserving!

Reply:
Dropping encodings through operations seems exceedingly dangerous and definitely semantically incorrect, at least for the intended use cases in my compiler. 

First, in the Tensor representation, an operation such as "SliceOp" should be producing a tensor that has the same encoding as the original tensor.

Second, if we allow operations to simply drop encodings, at least in my case, bufferization will either fail (since it cannot reconcile the types) or it will produce suboptimal code to convert the types using copies.

Is it possible that i'm completely misusing the encoding field? For our compiler it contains information on how the Tensor will be laid out in memory when allocated.

Reply:
You're basically saying we should do:

> 1. We don't do the transformation (because we risk making an incorrect transformation)

I don't disagree, I'm not sure we have a good documentation of the intent with respect to this tensor encoding.
(many folks, including me, were a bit concerned when this was introduced that we may be better introducing new types because this seems incredibly unsafe and it makes the codebase hard to get right: this is where we're at now)